src/Microsoft.AspNetCore.ResponseCompression/BodyWrapperStream.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.ResponseCompression
+{
+    /// <summary>
+    /// Stream wrapper that create specific compression stream only if necessary.
+    /// </summary>
+    internal class BodyWrapperStream : Stream
+    {
+        private readonly HttpResponse _response;
+
+        private readonly Stream _bodyOriginalStream;
+
+        private readonly HashSet<string> _mimeTypes;
+
+        private readonly IResponseCompressionProvider _compressionProvider;
+
+        private bool _compressionChecked = false;
+
+        private Stream _compressionStream = null;
+
+        internal BodyWrapperStream(HttpResponse response, Stream bodyOriginalStream, HashSet<string> mimeTypes, IResponseCompressionProvider compressionProvider)
```

Comment:
Maybe `ISet<string>` instead of `HashSet` if the unique elements are the only concern.